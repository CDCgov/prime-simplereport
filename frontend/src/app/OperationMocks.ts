import { WHOAMI_QUERY } from "./ReportingApp";

export const WhoAmIQueryMock = {
  request: {
    query: WHOAMI_QUERY,
    fetchPolicy: "no-cache",
  },
  result: {
    data: {
      whoami: {
        id: "05b2f71a-9392-442b-aab5-4eb550a864c0",
        firstName: "Bob",
        middleName: null,
        lastName: "Bobberoo",
        suffix: null,
        email: "bob@example.com",
        isAdmin: false,
        permissions: [
          "EDIT_PATIENT",
          "ARCHIVE_PATIENT",
          "READ_PATIENT_LIST",
          "EDIT_ORGANIZATION",
          "START_TEST",
          "EDIT_FACILITY",
          "ACCESS_ALL_FACILITIES",
          "READ_RESULT_LIST",
          "READ_ARCHIVED_PATIENT_LIST",
          "SUBMIT_TEST",
          "MANAGE_USERS",
          "SEARCH_PATIENTS",
          "UPDATE_TEST",
        ],
        roleDescription: "Admin user",
        organization: {
          name: "Dis Organization",
          testingFacility: [
            {
              id: "fec4de56-f4cc-4c61-b3d5-76869ca71296",
              name: "Testing Site",
            },
          ],
        },
      },
    },
  },
};
