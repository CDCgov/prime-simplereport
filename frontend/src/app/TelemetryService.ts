import {
  ApplicationInsights,
  ITelemetryItem,
  SeverityLevel,
} from "@microsoft/applicationinsights-web";
import { ReactPlugin } from "@microsoft/applicationinsights-react-js";

import {
  stripIdTokenFromOktaRedirectUri,
  stripIdTokenFromOperationName,
} from "./PrimeErrorBoundary";

let reactPlugin: ReactPlugin | null = null;
let appInsights: ApplicationInsights | null = null;

const createTelemetryService = () => {
  const initialize = () => {
    const connectionString =
      process.env.REACT_APP_APPLICATIONINSIGHTS_CONNECTION_STRING;

    if (!connectionString) {
      if (process.env.NODE_ENV !== "test") {
        console.warn("App Insights connection string not provided");
      }
      return;
    }

    reactPlugin = new ReactPlugin();

    appInsights = new ApplicationInsights({
      config: {
        connectionString,
        extensions: [reactPlugin],
        loggingLevelConsole: process.env.NODE_ENV === "development" ? 2 : 0,
        disableFetchTracking: false,
        enableAutoRouteTracking: true,
        loggingLevelTelemetry: 2,
        maxBatchInterval: 0,
      },
    });

    appInsights.addTelemetryInitializer(function (envelope: ITelemetryItem) {
      filterUnneededItems(envelope);
    });

    appInsights.loadAppInsights();
  };

  return { reactPlugin, appInsights, initialize };
};

export const ai = createTelemetryService();
export const getAppInsights = () => appInsights;

export function filterStaticFiles(envelope: ITelemetryItem) {
  try {
    const regexRemoteDependency =
      /Microsoft.ApplicationInsights.(.*).RemoteDependency/;

    const staticFilesToIgnore = [
      "GET /maintenance.json",
      "GET /app/static/commit.txt",
    ];

    if (
      regexRemoteDependency.test(envelope.name) &&
      staticFilesToIgnore.includes((envelope as any).baseData.name)
    ) {
      return false;
    }
  } catch (e) {
    /* do nothing and don't disrupt logging*/
  }
}

export function filterUnneededItems(envelope: ITelemetryItem) {
  const staticFileFound = filterStaticFiles(envelope) !== undefined;
  if (staticFileFound) return false;

  filterPotentialOktaRedirectEvent(envelope);
}

export function filterPotentialOktaRedirectEvent(envelope: ITelemetryItem) {
  // Okta redirects only come from page views events
  const eventIsPageView = envelope?.baseType === "PageviewData";
  if (!eventIsPageView) return true;

  const telemetryItem = envelope?.baseData;

  const telemetryItemNeedsIdSanitization =
    telemetryItem?.uri.includes("#id_token") ||
    envelope?.ext?.trace.name.includes("#id_token");

  if (
    telemetryItemNeedsIdSanitization &&
    telemetryItem?.uri &&
    envelope?.ext?.trace.name
  ) {
    const urlWithoutIdToken = stripIdTokenFromOktaRedirectUri(
      telemetryItem.uri
    );
    telemetryItem.uri = urlWithoutIdToken;

    // possible properties that need replacing
    if (telemetryItem?.properties?.refUri)
      telemetryItem.properties.refUri = urlWithoutIdToken;
    if (telemetryItem?.refURI) telemetryItem.refUri = urlWithoutIdToken;

    if (telemetryItem?.operation_Name) {
      telemetryItem.operation_Name = stripIdTokenFromOperationName(
        envelope!.ext!.trace.name
      );
    }
    return true;
  }
}

const logSeverityMap = {
  log: SeverityLevel.Information,
  warn: SeverityLevel.Warning,
  error: SeverityLevel.Error,
  info: SeverityLevel.Information,
} as const;

export function withInsights(console: Console) {
  const originalConsole = { ...console };

  Object.entries(logSeverityMap).forEach((el) => {
    const [method, severityLevel] = el as [
      keyof typeof logSeverityMap,
      SeverityLevel
    ];

    console[method] = (...data: any[]) => {
      originalConsole[method](...data);

      if (method === "error" || method === "warn") {
        let exception = data[0] instanceof Error ? data[0] : undefined;
        const id = (() => {
          let message = exception ? exception.message : data[0];
          if (typeof message === "string") {
            const messageNeedsSanitation = message.includes("#id_token");
            if (messageNeedsSanitation) {
              message = stripIdTokenFromOktaRedirectUri(message);
            }
          }
          if (exception) {
            exception = new Error(message);
            return message;
          }

          if (typeof data[0] === "string") {
            data[0] = message;
            return message;
          }
          const exceptionData = JSON.stringify(data[0]);
          return exceptionData;
        })();
        appInsights?.trackException({
          exception: exception,
          id,
          severityLevel,
          properties: {
            additionalInformation:
              data.length === 1 ? undefined : JSON.stringify(data.slice(1)),
          },
        });

        return;
      }

      const message =
        typeof data[0] === "string" ? data[0] : JSON.stringify(data[0]);
      appInsights?.trackEvent({
        name: `${method.toUpperCase()} - ${message}`,
        properties: {
          severityLevel,
          message,
          additionalInformation:
            data.length === 1 ? undefined : JSON.stringify(data.slice(1)),
        },
      });
    };
  });
}

export function getAppInsightsHeaders(): { [key: string]: string } {
  // x-ms-session-id is passed explicitly to the backend so we can correlate
  // backend operations with the frontend ones in App Insights
  return {
    "x-ms-session-id": getAppInsightsSessionId(),
  };
}

function getAppInsightsSessionId(): string {
  return appInsights?.context.getSessionId() ?? "";
}
