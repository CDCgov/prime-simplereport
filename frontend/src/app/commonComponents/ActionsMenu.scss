@use "../../scss/settings" as *;

.sr-actions-menu {
  color: $blue-cool-70;
  background: inherit;
  padding: 0;
  border: 0;
}

.sr-menu-item {
  margin-top: 0.25em;
  margin-bottom: 0.25em;
}
