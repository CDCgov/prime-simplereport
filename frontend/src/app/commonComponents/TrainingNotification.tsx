import { useEffect, useState } from "react";
import Modal from "react-modal";

import logo from "../../img/simplereport-logomark-color.svg";

import Alert from "./Alert";
import Button from "./Button/Button";

const SESSION_IDENTIFIER = "trainingAcknowledged";
export const TRAINING_PURPOSES_ONLY =
  "This is a training site with sample data for demonstration purposes only.";

const initialAcknowledged = !!sessionStorage.getItem(SESSION_IDENTIFIER);

export const TrainingNotification: React.FC = () => {
  const [acknowledged, setAcknowledged] = useState(initialAcknowledged);

  useEffect(() => {
    if (acknowledged) {
      sessionStorage.setItem(SESSION_IDENTIFIER, "yes");
    }
  }, [acknowledged]);

  return (
    <>
      <div className="usa-site-alert usa-site-alert--info usa-site-alert--slim border-top border-base-lighter">
        <Alert type="info" role="alert">
          {TRAINING_PURPOSES_ONLY}
        </Alert>
      </div>
      <Modal
        isOpen={!acknowledged}
        style={{
          content: {
            maxHeight: "90vh",
            width: "50em",
            position: "initial",
          },
        }}
        overlayClassName="prime-modal-overlay display-flex flex-align-center flex-justify-center"
        contentLabel="SimpleReport training site"
        ariaHideApp={process.env.NODE_ENV !== "test"}
      >
        <div className="sr-training-modal margin-1">
          <div className="display-flex flex-align-start">
            <img
              src={logo}
              alt="SimpleReport logo"
              className="width-10 margin-right-3 margin-top-1"
            />
            <div className="usa-prose">
              <h1 className="font-heading-xl text-normal">
                Welcome to the SimpleReport training site!
              </h1>
              <h2 className="font-heading-lg text-normal margin-top-105">
                A few important reminders:
              </h2>
              <ul>
                <li>{TRAINING_PURPOSES_ONLY}</li>
                <li>
                  Don't enter personally identifiable information (PII) or
                  protected health information (PHI) on this site.
                </li>
                <li>
                  Results entered on this training site will not be reported.
                </li>
              </ul>
              <Button
                onClick={() => {
                  setAcknowledged(true);
                }}
                className="margin-right-0"
              >
                Got it, thanks
              </Button>
            </div>
          </div>
        </div>
      </Modal>
    </>
  );
};
