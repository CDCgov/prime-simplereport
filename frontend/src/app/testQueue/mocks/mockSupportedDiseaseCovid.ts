import mockSupportedDiseaseTestPerformedCovid from "../../supportAdmin/DeviceType/mocks/mockSupportedDiseaseTestPerformedCovid";

const mockSupportedDiseaseCovid = [
  {
    supportedDisease:
      mockSupportedDiseaseTestPerformedCovid[0].supportedDisease,
    testPerformedLoincCode: "123456",
  },
];

export default mockSupportedDiseaseCovid;
