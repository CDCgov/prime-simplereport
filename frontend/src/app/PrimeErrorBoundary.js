import React from "react";

// import { SeverityLevel } from "@microsoft/applicationinsights-web";
import ErrorPage from "./commonComponents/ErrorPage";
import { getAppInsights } from "./TelemetryService";

export default class PrimeErrorBoundary extends React.Component {
  state = { hasError: false, error: null };

  componentDidCatch(error, info) {
    this.setState({ hasError: true, error });
    const appInsights = getAppInsights();
    if (appInsights) {
      getAppInsights().trackException({
        exception: new Error("PrimeErrorBoundary: " + error),
        severityLevel: 3,
        properties: { componentStack: info.componentStack },
      });
    }
  }

  render() {
    if (this.state.hasError) {
      return React.createElement(ErrorPage);
    }

    return this.props.children;
  }
}
