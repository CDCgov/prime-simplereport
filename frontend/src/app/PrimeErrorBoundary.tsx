import React, { ErrorInfo } from "react";

import ErrorPage from "./commonComponents/ErrorPage";
import { getAppInsights } from "./TelemetryService";
import { getUrl } from "./utils/url";

interface ErrorBoundaryState {
  hasError: boolean;
  redirectToOkta: boolean;
}

interface PrimeErrorBoundaryProps {
  children?: React.ReactNode;
}

export default class PrimeErrorBoundary extends React.Component<
  PrimeErrorBoundaryProps,
  ErrorBoundaryState
> {
  constructor(props: any) {
    super(props);
    this.state = { hasError: false, redirectToOkta: false };
  }

  static getDerivedStateFromError(error: Error) {
    if (error.message === "Not authenticated, redirecting to Okta...") {
      return { redirectToOkta: true };
    }
    // Update state so the next render will show the fallback UI.
    return { hasError: true };
  }

  componentDidCatch(error: Error, info: ErrorInfo) {
    if (error.message === "Not authenticated, redirecting to Okta...") {
      return;
    }

    const appInsights = getAppInsights();
    if (appInsights) {
      appInsights.trackException({
        exception: new Error("PrimeErrorBoundary: " + error),
        severityLevel: 3,
        properties: { componentStack: info.componentStack },
      });
    }
  }

  render() {
    if (this.state.hasError) {
      return <ErrorPage />;
    }

    if (this.state.redirectToOkta) {
      const url = `${process.env.REACT_APP_OKTA_URL}/oauth2/default/v1/authorize`;
      const clientId = `?client_id=${process.env.REACT_APP_OKTA_CLIENT_ID}`;
      const redirectUri = `&redirect_uri=${encodeURIComponent(getUrl())}`;
      const responseType = `&response_type=${encodeURIComponent(
        "token id_token"
      )}`;
      const scope =
        "&scope=" +
        encodeURIComponent(
          `openid simple_report ${process.env.REACT_APP_OKTA_SCOPE}`
        );
      const nonce = "&nonce=thisisnotsafe";
      const state = "&state=thisisbogus";
      window.location.href =
        url + clientId + redirectUri + responseType + scope + nonce + state;
      return false;
    }

    return this.props.children;
  }
}

export function stripIdTokenFromOktaRedirectUri(uri: string) {
  const regexJWTAsQueryParam = /(?<=#id_token=).*?(?=&token_type=)/;
  return stripIdTokenFromString(regexJWTAsQueryParam, uri);
}

export function stripIdTokenFromOperationName(operationName: string) {
  const regexOperationName = /(?<=#id_token=).*/;
  return stripIdTokenFromString(regexOperationName, operationName);
}

function stripIdTokenFromString(regex: RegExp, string: string) {
  const idTokenFound = string.match(regex);
  if (idTokenFound === null) return string;
  return string.replace(idTokenFound[0], "{ID-TOKEN-OBSCURED}");
}
