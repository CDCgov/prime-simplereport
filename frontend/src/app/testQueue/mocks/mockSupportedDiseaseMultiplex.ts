import mockSupportedDiseaseTestPerformedMultiplex from "../../supportAdmin/DeviceType/mocks/mockSupportedDiseaseTestPerformedMultiplex";

import mockSupportedDiseaseCovid from "./mockSupportedDiseaseCovid";

export const mockSupportedDiseaseFlu = [
  {
    supportedDisease:
      mockSupportedDiseaseTestPerformedMultiplex[1].supportedDisease,
    testPerformedLoincCode: "456789",
  },
  {
    supportedDisease:
      mockSupportedDiseaseTestPerformedMultiplex[2].supportedDisease,
    testPerformedLoincCode: "789123",
  },
];

const mockSupportedDiseaseMultiplex = [
  ...mockSupportedDiseaseCovid,
  ...mockSupportedDiseaseFlu,
];

export default mockSupportedDiseaseMultiplex;
