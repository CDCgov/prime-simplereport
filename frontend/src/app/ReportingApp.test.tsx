import { MemoryRouter, Route, Routes } from "react-router-dom";
import { Provider } from "react-redux";
import createMockStore, { MockStoreEnhanced } from "redux-mock-store";
import { MockedProvider, MockedResponse } from "@apollo/client/testing";
import {
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { ApplicationInsights } from "@microsoft/applicationinsights-web";
import jwtDecode from "jwt-decode";
import MockDate from "mockdate";

import {
  GetFacilityQueueDocument,
  GetFacilityQueueQuery,
  GetTopLevelDashboardMetricsNewDocument,
} from "../generated/graphql";

import ReportingApp, { WHOAMI_QUERY } from "./ReportingApp";
import PrimeErrorBoundary from "./PrimeErrorBoundary";
import { TRAINING_PURPOSES_ONLY } from "./commonComponents/TrainingNotification";
import {
  getEndDateFromDaysAgo,
  getStartDateFromDaysAgo,
} from "./analytics/Analytics";
import { getAppInsights } from "./TelemetryService";
import mockSupportedDiseaseMultiplex from "./testQueue/mocks/mockSupportedDiseaseMultiplex";
import mockSupportedDiseaseCovid from "./testQueue/mocks/mockSupportedDiseaseCovid";
import { WhoAmIQueryMock } from "./OperationMocks";

const mockDispatch = jest.fn();

jest.mock("uuid");
jest.mock("./VersionService", () => ({
  VersionService: {
    enforce: jest.fn(),
  },
}));
jest.mock("./testResults/viewResults/CleanTestResultsList", () => {
  return () => <p>CleanTestResultsList</p>;
});
jest.mock("./testResults/viewResults/TestResultsList", () => {
  return () => <p>TestResultsList</p>;
});
jest.mock("./TelemetryService", () => ({
  ...jest.requireActual("./TelemetryService"),
  getAppInsights: jest.fn(),
}));
jest.mock("jwt-decode", () => jest.fn());

const mockStore = createMockStore([]);

const store = {
  dispatch: mockDispatch,
  organization: {
    name: "Organization Name",
  },
  user: {
    id: "05b2f71a-9392-442b-aab5-4eb550a864c0",
    firstName: "Bob",
    middleName: null,
    lastName: "Bobberoo",
    suffix: null,
    email: "bob@example.com",
    roleDescription: "Admin user",
    isAdmin: false,
    permissions: [
      "EDIT_PATIENT",
      "ARCHIVE_PATIENT",
      "READ_PATIENT_LIST",
      "EDIT_ORGANIZATION",
      "START_TEST",
      "EDIT_FACILITY",
      "ACCESS_ALL_FACILITIES",
      "READ_RESULT_LIST",
      "READ_ARCHIVED_PATIENT_LIST",
      "SUBMIT_TEST",
      "MANAGE_USERS",
      "SEARCH_PATIENTS",
      "UPDATE_TEST",
    ],
  },
  facilities: [
    {
      id: "fec4de56-f4cc-4c61-b3d5-76869ca71296",
      name: "Testing Site",
    },
  ],
  facility: {
    id: "fec4de56-f4cc-4c61-b3d5-76869ca71296",
    name: "Testing Site",
  },
};

store.dispatch = mockDispatch;

const facilityQueryMock = {
  request: {
    query: GetFacilityQueueDocument,
    fetchPolicy: "no-cache",
    variables: {
      facilityId: "fec4de56-f4cc-4c61-b3d5-76869ca71296",
    },
  },
  result: {
    data: {
      queue: [],
      facility: {
        id: "fec4de56-f4cc-4c61-b3d5-76869ca71296",
        name: "Testing site",
        deviceTypes: [
          {
            internalId: "d70bb3b8-96bd-40d9-a3ce-b266a7edb91d",
            name: "Quidel Sofia 2",
            testLength: 15,
            supportedDiseaseTestPerformed: mockSupportedDiseaseMultiplex,
            swabTypes: [],
          },
          {
            internalId: "5e44dcef-8cc6-44f4-a200-a5b8169ab60a",
            name: "LumiraDX",
            testLength: 15,
            supportedDiseaseTestPerformed: mockSupportedDiseaseCovid,
            swabTypes: [],
          },
        ],
      },
    } as GetFacilityQueueQuery,
  },
};
const WhoAmIErrorQueryMock = {
  request: {
    query: WHOAMI_QUERY,
    fetchPolicy: "no-cache",
  },
  error: new Error("Server connection error"),
};
const getAnalyticsQueryMock = () => ({
  request: {
    query: GetTopLevelDashboardMetricsNewDocument,
    variables: {
      facilityId: "",
      startDate: getStartDateFromDaysAgo(7),
      endDate: getEndDateFromDaysAgo(0),
      disease: "COVID-19",
    },
  },
  result: {
    data: {
      topLevelDashboardMetrics: {
        totalTestCount: 124820,
        positiveTestCount: 2270,
      },
    },
  },
});

const renderApp = (
  newStore: MockStoreEnhanced<unknown, {}>,
  queryMocks: MockedResponse[]
) => {
  return {
    user: userEvent.setup(),
    ...render(
      <PrimeErrorBoundary>
        <Provider store={newStore}>
          <MockedProvider mocks={queryMocks} addTypename={false}>
            <MemoryRouter>
              <Routes>
                <Route path="/*" element={<ReportingApp />} />
              </Routes>
            </MemoryRouter>
          </MockedProvider>
        </Provider>
      </PrimeErrorBoundary>
    ),
  };
};

const MODAL_TEXT = "Welcome to the SimpleReport";

describe("App", () => {
  beforeEach(() => {
    MockDate.set("2021-08-01");
  });

  afterEach(() => {
    MockDate.reset();
  });

  it("Renders loading masks (account and facility)", async () => {
    const mockedStore = mockStore({ ...store });
    renderApp(mockedStore, [WhoAmIQueryMock, facilityQueryMock]);
    await screen.findByText("Loading account information...");
    expect(await screen.findByText("Loading facility information..."));
  });

  it("Render main screen", async () => {
    const mockedStore = mockStore({ ...store, dataLoaded: true });
    const { user } = renderApp(mockedStore, [
      WhoAmIQueryMock,
      facilityQueryMock,
      getAnalyticsQueryMock(),
    ]);

    await waitForElementToBeRemoved(() =>
      screen.queryByText("Loading account information...")
    );

    await user.click(screen.getAllByText("Testing Site", { exact: false })[0]);
    expect(
      await screen.findByText("COVID-19 testing data", { exact: false })
    ).toBeInTheDocument();
  });

  it("should show error UI", async () => {
    const mockedStore = mockStore({ ...store, dataLoaded: true });
    const { container } = renderApp(mockedStore, [WhoAmIErrorQueryMock]);
    await screen.findByText("error", { exact: false });
    expect(container).toMatchSnapshot();
  });

  it("displays the training header and modal and dismisses the modal", async () => {
    process.env.REACT_APP_IS_TRAINING_SITE = "true";
    const mockedStore = mockStore({ ...store, dataLoaded: true });
    const { user } = renderApp(mockedStore, [
      WhoAmIQueryMock,
      facilityQueryMock,
      getAnalyticsQueryMock(),
    ]);
    expect(await screen.findAllByText(TRAINING_PURPOSES_ONLY)).toHaveLength(2);
    const trainingWelcome = await screen.findByText(MODAL_TEXT, {
      exact: false,
    });
    expect(trainingWelcome).toBeInTheDocument();

    await user.click(
      screen.getByText(/got it,? thanks/i, { selector: "button" })
    );
    expect(trainingWelcome).not.toBeInTheDocument();
  });

  it("does not display training notifications outside the training environment", () => {
    process.env.REACT_APP_IS_TRAINING_SITE = "false";
    const mockedStore = mockStore({ ...store, dataLoaded: true });
    renderApp(mockedStore, [WhoAmIQueryMock, facilityQueryMock]);
    expect(screen.queryByText(TRAINING_PURPOSES_ONLY)).not.toBeInTheDocument();
    expect(screen.queryByText(MODAL_TEXT)).not.toBeInTheDocument();
  });

  describe("logs to App Insights on WhoAmI error", () => {
    const oldTokenClaim = process.env.REACT_APP_OKTA_TOKEN_ROLE_CLAIM;
    const trackExceptionMock = jest.fn();

    beforeEach(() => {
      process.env.REACT_APP_OKTA_TOKEN_ROLE_CLAIM = "test_roles";
      trackExceptionMock.mockReset();
      (getAppInsights as jest.Mock).mockImplementation(() => {
        const ai = Object.create(ApplicationInsights.prototype);
        return Object.assign(ai, { trackException: trackExceptionMock });
      });
    });
    afterEach(() => {
      process.env.REACT_APP_OKTA_TOKEN_ROLE_CLAIM = oldTokenClaim;
      jest.spyOn(Storage.prototype, "getItem").mockRestore();
    });
    it("logs with access token info", async () => {
      jest
        .spyOn(Storage.prototype, "getItem")
        .mockImplementation(() => "definitely a valid token");
      (jwtDecode as jest.Mock).mockReturnValue({
        sub: "subject@fakeorg.net",
        test_roles: [
          "SR-FAKE-TENANT:XX-TestOrg-123:NO_ACCESS",
          "SR-FAKE-TENANT:XX-TestOrg-123:USER",
        ],
      });
      const mockedStore = mockStore({ ...store, dataLoaded: true });

      renderApp(mockedStore, [WhoAmIErrorQueryMock]);
      await screen.findByText("error", { exact: false });

      expect(trackExceptionMock).toHaveBeenCalledWith({
        exception: new Error("Server connection error"),
        properties: {
          "user message": "Server connection error",
          "valid access token": true,
          "token subject": "subject@fakeorg.net",
          "token roles": [
            "SR-FAKE-TENANT:XX-TestOrg-123:NO_ACCESS",
            "SR-FAKE-TENANT:XX-TestOrg-123:USER",
          ],
        },
      });
    });
    it("still logs if missing token", async () => {
      const mockedStore = mockStore({ ...store, dataLoaded: true });
      jest.spyOn(Storage.prototype, "getItem").mockImplementation(() => null);

      renderApp(mockedStore, [WhoAmIErrorQueryMock]);
      await screen.findByText("error", { exact: false });

      expect(trackExceptionMock).toHaveBeenCalledWith({
        exception: new Error("Server connection error"),
        properties: {
          "user message": "Server connection error",
          "valid access token": null,
          "token subject": undefined,
          "token roles": undefined,
        },
      });
    });
    it("still logs if invalid token", async () => {
      jest
        .spyOn(Storage.prototype, "getItem")
        .mockImplementation(() => "definitely NOT a valid token");
      jest.spyOn(console, "error").mockImplementation(() => {});
      (jwtDecode as jest.Mock).mockImplementation(() => {
        throw new Error("InvalidToken");
      });
      const mockedStore = mockStore({ ...store, dataLoaded: true });

      renderApp(mockedStore, [WhoAmIErrorQueryMock]);
      await screen.findByText("error", { exact: false });

      expect(trackExceptionMock).toHaveBeenCalledWith({
        exception: new Error("Server connection error"),
        properties: {
          "user message": "Server connection error",
          "valid access token": false,
          "token subject": undefined,
          "token roles": undefined,
        },
      });
    });
  });
  it("renders CleanTestResultsList when going to /results/", async () => {
    const trackMetricMock = jest.fn();
    const setAuthenticatedUserContextMock = jest.fn();
    Object.defineProperty(global, "visualViewport", {
      value: { width: 1200, height: 800 },
    });
    (getAppInsights as jest.Mock).mockImplementation(() => {
      const ai = Object.create(ApplicationInsights.prototype);
      return Object.assign(ai, {
        trackMetric: trackMetricMock,
        setAuthenticatedUserContext: setAuthenticatedUserContextMock,
      });
    });

    const mockedStore = mockStore({ ...store, dataLoaded: true });

    render(
      <Provider store={mockedStore}>
        <MockedProvider mocks={[WhoAmIQueryMock]} addTypename={false}>
          <MemoryRouter
            initialEntries={[
              `/results?facility=fec4de56-f4cc-4c61-b3d5-76869ca71296`,
            ]}
          >
            <ReportingApp />
          </MemoryRouter>
        </MockedProvider>
      </Provider>
    );

    expect(await screen.findByText("CleanTestResultsList")).toBeInTheDocument();
    await waitFor(() =>
      expect(trackMetricMock).toHaveBeenCalledWith(
        {
          name: "userViewport_reporting",
          average: 1200,
        },
        {
          width: 1200,
          height: 800,
        }
      )
    );

    expect(setAuthenticatedUserContextMock).toHaveBeenCalled();
  });

  it("renders TestResultsList when going to /results/page_number ", async () => {
    const mockedStore = mockStore({ ...store, dataLoaded: true });

    render(
      <Provider store={mockedStore}>
        <MockedProvider mocks={[WhoAmIQueryMock]} addTypename={false}>
          <MemoryRouter
            initialEntries={[
              `/results/1?facility=fec4de56-f4cc-4c61-b3d5-76869ca71296`,
            ]}
          >
            <ReportingApp />
          </MemoryRouter>
        </MockedProvider>
      </Provider>
    );
    expect(await screen.findByText("TestResultsList")).toBeInTheDocument();
  });
});
