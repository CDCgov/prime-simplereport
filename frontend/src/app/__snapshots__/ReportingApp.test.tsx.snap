// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should show error UI 1`] = `
<div>
  <p>
    Server connection error...
  </p>
</div>
`;
